crates/sparrow-execution/src/lib.rs

```diff
@@ -8,5 +8,228 @@
+    ) -> error_stack::Result<Self, Error> {
+        let mut executor = PlanExecutor {
+            worker_pool: WorkerPool::new(query_id).change_context(Error::Creating)?,
+            source_tasks: SourceTasks::default(),
+            read_config: Arc::new(ReadConfig {
+                keep_open: false,
+                start_time: None,
+                end_time: None,
+            }),
+        };
+
+        let last_step = plan.steps.last().expect("at least one step");
+        let output_schema = result_type_to_output_schema(&last_step.result_type);
+
+        let sink_pipeline = executor
+            .worker_pool
+            .add_pipeline(1, WriteChannelPipeline::new(output, output_schema));
+
+        // Map from the producing step ID to the consumers.
+        let mut step_consumers: HashMap<StepId, InputHandles> = HashMap::new();
+        step_consumers
+            .entry(plan.steps.last_idx())
+            .or_default()
+            .add_consumer(sink_pipeline, 0);
+
+        // Iterate in reverse so the receivers (consumers) are created before the senders (producers).
+        for pipeline in plan.pipelines.iter().rev() {
+            let first_step_id = *pipeline.steps.first().expect("at least one step");
+            let last_step_id = *pipeline.steps.last().expect("at least one step");
+
+            // Get the consumer(s) for the last step in the pipeline.
+            // While we schedule and execute pipelines, the structure of the individual
+            // steps identify which *step* they receive from. So, to determine where
+            // this pipeline "pushes output" we need to determine where the last step
+            // in this pipeline "pushes output".
+            let consumers = step_consumers
+                .remove(&last_step_id)
+                .expect("at least one consumer for step");
+
+            let first_step_inputs = &plan.steps[first_step_id].inputs;
+
+            // Create a transform pipeline (if possible), othrewise create the
+            // appropriate non-transform pipeline.
+            let pipeline = if pipeline
+                .steps
+                .iter()
+                .all(|id| plan.steps[*id].kind.is_transform())
+            {
+                debug_assert_eq!(
+                    first_step_inputs.len(),
+                    1,
+                    "Transforms should have a single input"
+                );
+                let first_step_input_id = first_step_inputs[0];
+
+                // If all of the steps are transforms, then we use the transform pipeline.
+                executor.add_transform_pipeline(
+                    first_step_input_id,
+                    &plan,
+                    &pipeline.steps,
+                    consumers,
+                )?
+            } else {
+                assert_eq!(
+                    pipeline.steps.len(),
+                    1,
+                    "Non-transforms should be in separate pipelines"
```

Comment:
Would "Non-transforms should have a single step" be more descriptive here?

Reply:
I don't think so. The error here is that right now, we can't execute multiple non-transform steps as part of a single pipeline, since each of those non-transform steps is a separate pipelien. Given that the *step* is a transform or non-transform, it seems confusing to say "non transforms should have a single step"... every transform step is a single step as is every non-transform step.

Reply:
Ah, right. I'm still getting the grasp of the step to pipeline conversion. 

Can you add "step" to the error to help remind me? 

`Non-transform steps should be in (or, be executed) as separated pipelines`

Reply:
Yep. Done.